File: src/18.4Sum.kt
fun main() {
    println(fourSum(intArrayOf(1, 0, -1, 0, -2, 2), 0).joinToString())
    println(fourSum(intArrayOf(2, 2, 2, 2, 2), 8).joinToString())
    println(fourSum(intArrayOf(0, 0, 0, 0), 0).joinToString())
    println(
            fourSum(intArrayOf(1000000000, 1000000000, 1000000000, 1000000000), -294967296)
                    .joinToString()
    )
}

fun fourSum(nums: IntArray, target: Int): List<List<Int>> {
    nums.sort()

    val set = mutableSetOf<List<Int>>()

    for (i in 0..nums.size - 3) {
        for (j in i + 1..nums.size - 2) {
            val value = target.toLong() - nums[i] - nums[j]
            var left = j + 1
            var right = nums.size - 1

            while (left < right) {
                val sum = nums[left].toLong() + nums[right]

                if (value == sum) {
                    set.add(listOf(nums[i], nums[j], nums[left], nums[right]))
                    left++
                    right--
                } else if (value > sum) {
                    left++
                } else {
                    right--
                }
            }
        }
    }

    return set.toList()
}

File: lib/18.4Sum.dart
void main() {
  print(fourSum([1, 0, -1, 0, -2, 2], 0).toString());
  print(fourSum([2, 2, 2, 2, 2], 8).toString());
  print(fourSum([0, 0, 0, 0], 0).toString());
  print(fourSum([1000000000, 1000000000, 1000000000, 1000000000], -294967296)
      .toString());
}

List<List<int>> fourSum(List<int> nums, int target) {
  nums.sort();

  List<List<int>> list = [];

  for (var i = 0; i < nums.length - 3; i++) {
    if (i != 0 && nums[i] == nums[i - 1]) {
      continue;
    }

    for (var j = i + 1; j < nums.length - 2; j++) {
      if (j > i + 1 && nums[j] == nums[j - 1]) {
        continue;
      }

      final value = target - nums[i] - nums[j];
      var left = j + 1;
      var right = nums.length - 1;

      while (left < right) {
        final sum = nums[left] + nums[right];

        if (value == sum) {
          list.add([nums[i], nums[j], nums[left], nums[right]]);

          // Skip duplicates
          while (left < right && nums[left] == nums[left + 1]) left++;
          while (left < right && nums[right] == nums[right - 1]) right--;

          left++;
          right--;
        } else if (value > sum) {
          left++;
        } else {
          right--;
        }
      }
    }
  }

  return list;
}
